pkg/async/notifications/implementations/cloudevent_publisher.go

```diff
@@ -0,0 +1,133 @@
+package implementations
+
+import (
+	"context"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/cloudevents/sdk-go/protocol/kafka_sarama/v2"
+
+	"github.com/NYTimes/gizmo/pubsub"
+	pbcloudevents "github.com/cloudevents/sdk-go/binding/format/protobuf/v2"
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/google/uuid"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/flyteorg/flyteadmin/pkg/async/notifications/interfaces"
+
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/flyteorg/flytestdlib/promutils"
+	"github.com/golang/protobuf/proto"
+)
+
+const (
+	cloudEventSource = "https://github.com/flyteorg/flyteadmin"
+	jsonSchemaURL    = "jsonschemaurl"
+)
+
+type Receiver = string
+
+const (
+	Kafka Receiver = "Kafka"
+)
+
+type CloudEventSender interface {
+	Send(ctx context.Context, notificationType string, event cloudevents.Event) error
+}
+
+type PubSubSender struct {
+	Pub pubsub.Publisher
+}
+
+func (s *PubSubSender) Send(ctx context.Context, notificationType string, event cloudevents.Event) error {
+	eventByte, err := pbcloudevents.Protobuf.Marshal(&event)
+	if err != nil {
+		logger.Errorf(ctx, "Failed to marshal cloudevent with error: %v", err)
+		return err
+	}
+	if err := s.Pub.PublishRaw(ctx, notificationType, eventByte); err != nil {
+		logger.Errorf(ctx, "Failed to publish a message with key [%s] and message [%s] and error: %v", notificationType, event.String(), err)
+		return err
+	}
+
+	return nil
+}
+
+type KafkaSender struct {
+	Client cloudevents.Client
+}
+
+func (s *KafkaSender) Send(ctx context.Context, notificationType string, event cloudevents.Event) error {
+	if result := s.Client.Send(
+		// Set the producer message key
+		kafka_sarama.WithMessageKey(context.Background(), sarama.StringEncoder(event.ID())),
+		event,
+	); cloudevents.IsUndelivered(result) {
+		logger.Errorf(ctx, "failed to send: %v", result)
+	}
+	return nil
+}
+
+type CloudEventPublisher struct {
+	sender        CloudEventSender
+	systemMetrics eventPublisherSystemMetrics
+	events        sets.String
+}
+
+func (p *CloudEventPublisher) Publish(ctx context.Context, notificationType string, msg proto.Message) error {
+	p.systemMetrics.PublishTotal.Inc()
+
+	if !p.shouldPublishEvent(notificationType) {
+		return nil
+	}
+	logger.Debugf(ctx, "Publishing the following message [%+v]", msg)
+
+	event := cloudevents.NewEvent()
+	// CloudEvent specification: https://github.com/cloudevents/spec/blob/v1.0/spec.md#required-attributes
+	event.SetType(notificationType)
+	event.SetSource(cloudEventSource)
+	event.SetID(uuid.New().String())
+	event.SetTime(time.Now())
+	event.SetExtension(jsonSchemaURL, "https://github.com/pingsutw/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json")
```

Comment:
is this final? are we checking this into idl?

Reply:
sorry, the link is outdated. The new schema url is https://github.com/flyteorg/flyteidl/blob/cloudevent2/jsonschema/workflow_execution.json